Hazel/src/Hazel/ImGui/ImGuiConsole.cpp

```diff
@@ -0,0 +1,249 @@
+
+		// Buttons to quickly change level right
+		if (ImGui::ArrowButton("##MessageRenderFilter_R", ImGuiDir_Right))
+		{
+			s_MessageBufferRenderFilter = Message::GetHigherLevel(s_MessageBufferRenderFilter);
+		}
+
+		ImGui::SameLine(0.0f, spacing);
+
+		// Button for advanced settings
+		ImGui::SameLine(0.0f, ImGui::GetContentRegionAvail().x - ImGui::CalcTextSize("Settings").x - style.WindowPadding.x / 2.0f);
+		if (ImGui::Button("Settings"))
+			ImGui::OpenPopup("SettingsPopup");
+		if (ImGui::BeginPopup("SettingsPopup"))
+		{
+			ImGuiRenderSettings();
+			ImGui::EndPopup();
+		}
+
+	}
+
+	void ImGuiConsole::ImGuiRendering::ImGuiRenderSettings()
+	{
+		const float maxWidth = ImGui::CalcTextSize("Scroll to bottom").x * 1.1f;
+		const float spacing = ImGui::GetStyle().ItemInnerSpacing.x + ImGui::CalcTextSize(" ").x;
+
+		// Checkbox for scrolling lock
+		ImGui::AlignFirstTextHeightToWidgets();
+		ImGui::Text("Scroll to bottom");
+		ImGui::SameLine(0.0f, spacing + maxWidth - ImGui::CalcTextSize("Scroll to bottom").x);
+		ImGui::Checkbox("##ScrollToBottom", &s_AllowScrollingToBottom);
+
+		ImGui::SameLine(0.0f, spacing);
+
+		// Button to clear the console
+		ImGui::SameLine(0.0f, ImGui::GetContentRegionAvail().x - ImGui::CalcTextSize("Clear console").x + 1.0f);
+		if (ImGui::Button("Clear console"))
+			ImGuiConsole::Clear();
+
+		// Slider for font scale
+		ImGui::AlignFirstTextHeightToWidgets();
+		ImGui::Text("Display scale");
+		ImGui::SameLine(0.0f, spacing + maxWidth - ImGui::CalcTextSize("Display scale").x);
+		ImGui::PushItemWidth(maxWidth * 1.25f / 1.1f);
+		ImGui::SliderFloat("##DisplayScale", &s_DisplayScale, 0.5f, 4.0f, "%.1f");
+		ImGui::PopItemWidth();
+	}
+
+	void ImGuiConsole::ImGuiRendering::ImGuiRenderMessages()
+	{
+
+		ImGui::BeginChild("ScrollRegion", ImVec2(0, 0), false, ImGuiWindowFlags_HorizontalScrollbar);
+		{
+			ImGui::SetWindowFontScale(s_DisplayScale);
+
+			auto messageStart = s_MessageBuffer.begin() + s_MessageBufferBegin;
+			if (*messageStart) // If contains old message here
+				for (auto message = messageStart; message != s_MessageBuffer.end(); message++)
+					(*message)->OnImGuiRender();
+			if (s_MessageBufferBegin != 0) // Skipped first messages in vector
+				for (auto message = s_MessageBuffer.begin(); message != messageStart; message++)
+					(*message)->OnImGuiRender();
+
+			if (s_RequestScrollToBottom && ImGui::GetScrollMaxY() > 0)
+			{
+				ImGui::SetScrollY(ImGui::GetScrollMaxY());
+				s_RequestScrollToBottom = false;
+			}
+		}
+		ImGui::EndChild();
+	}
+
+	std::vector<ImGuiConsole::Message::Level> ImGuiConsole::Message::s_Levels{
+		ImGuiConsole::Message::Level::Trace,
+		//ImGuiConsole::Message::Level::Debug,
+		ImGuiConsole::Message::Level::Info,
+		ImGuiConsole::Message::Level::Warn,
+		ImGuiConsole::Message::Level::Error,
+		ImGuiConsole::Message::Level::Critical,
+		ImGuiConsole::Message::Level::Off
+	};
+
+	ImGuiConsole::Message::Message(const std::string message, Level level)
```

Comment:
Was this meant to be 'const std::string &' ? This potentially should a standard std::string then you can std::move it into m_Message